python/pyarrow/array.pxi

```diff
@@ -549,6 +550,11 @@ cdef class _PandasConvertible:
             Cast integers with nulls to objects
         date_as_object : bool, default True
             Cast dates to objects. If False, convert to datetime64[ns] dtype.
+        timestamp_as_object : bool, default False
+            Cast non-nanosecond timestamps (np.datetime64) to objects. This is
```

Comment:
Shoudn't we also force this conversion when you have nanosecond timestamps? If I specify `timestamp_as_object=True`, I would expect to actually get datetime objects, even if my unit is nanoseconds?

Reply:
Hm. It's lossy if you do that, though, since `datetime` only supports microseconds. And the goal is more to deal with out-of-range timestamps, and returning object is more a side-effect of trying to do that.

Actually, here's the argument against changing it: having multiple columns, some nanosecond, some not. If nanosecond gets downgraded to microsecond, you're forced to choice between losing data and not being able to access data. Whereas current default allows both accuracy and range.

I'm not sure in practice how much that will occur.

I can change it if you have a strong opinion in favor of making it always return objects.

Reply:
Hmm, yes, that's a good reason. 
Personally I don't think that it being lossy is a problem (in the end, the user would explicitly ask for it), but since this keyword applies to all columns, it might indeed be useful to handle this case of mixed-unit columns a bit more gracefully.

(this might also signal that a simple boolean flag is maybe not sufficient for this keyword)